### PR TITLE
WIP - Use envoy v3 API 

### PR DESCRIPTION
**Pull request by jmprusi**
# Changes

- Uses Envoy v3 API instead of the deprecated v2.

/kind enhancement

**Release Note**

```release-note
Use envoy v3 API instead of the deprecated v2 version. 
```

### CODE REVIEW

File: config/300-gateway.yaml [hunk 1]
```diff
@@ -36,7 +36,7 @@ spec:
             - --log-level info
           command:
             - /usr/local/bin/envoy
-          image: docker.io/envoyproxy/envoy:v1.16-latest
+          image: docker.io/envoyproxy/envoy:v1.17-latest
```
Comment (markusthoemmes) [inline config/300-gateway.yaml:39]: I'd leave the default at 1.16 for now, to make sure upgrade cases will go just fine. We can test 1.17 in addition to that though via the gateway list above.

File: pkg/config/ext_authz.go [hunk 2]
```diff
@@ -23,11 +23,11 @@ import (
 	"strconv"
 	"time"
 
-	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
-	endpoint "github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
-	extAuthService "github.com/envoyproxy/go-control-plane/envoy/config/filter/http/ext_authz/v2"
-	httpconnectionmanagerv2 "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	v3Cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```
Comment (markusthoemmes) [inline pkg/config/ext_authz.go:26]: nit: Let's just make this `cluster`

File: pkg/envoy/api/cluster.go [hunk 3]
```diff
@@ -19,27 +19,29 @@ package envoy
 import (
 	"time"
 
-	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
-	endpoint "github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
+	envoyCluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```
Comment (markusthoemmes) [inline pkg/envoy/api/cluster.go:22]:
```suggestion
	cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```

File: pkg/envoy/api/cluster_test.go [hunk 4]
```diff
@@ -20,8 +20,8 @@ import (
 	"testing"
 	"time"
 
-	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	endpoint "github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
+	v3Cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```
Comment (markusthoemmes) [inline pkg/envoy/api/cluster_test.go:23]:
```suggestion
	cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```

File: pkg/envoy/api/http_connection_manager.go [hunk 5]
```diff
@@ -31,35 +30,35 @@ import (
...
+	filters = append(filters, &hcm.HttpFilter{
 		Name: wellknown.Router,
 	})
-
 	// Write access logs to stdout by default.
-	accessLog, _ := ptypes.MarshalAny(&accesslog_v2.FileAccessLog{
-		Path: "/dev/stdout",
+	accessLog, _ := ptypes.MarshalAny(&fileaccesslog.FileAccessLog{
+		Path:            "/dev/stdout",
+		AccessLogFormat: nil,
```
Comment (markusthoemmes) [inline pkg/envoy/api/http_connection_manager.go:47]: Is this change necessary?

File: pkg/envoy/api/lb_endpoint_test.go [hunk 6]
```diff
@@ -19,8 +19,8 @@ package envoy
 import (
 	"testing"
 
-	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
-	envoy_api_v2_endpoint "github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3Endpoint "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
```
Comment (markusthoemmes) [inline pkg/envoy/api/lb_endpoint_test.go:23]:
```suggestion
	endpoint "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
```

File: pkg/envoy/api/listener_test.go [hunk 7]
```diff
@@ -21,10 +21,10 @@ import (
 	"reflect"
 	"testing"
 
-	envoy_api_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	auth "github.com/envoyproxy/go-control-plane/envoy/api/v2/auth"
-	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
-	envoy_api_v2_listener "github.com/envoyproxy/go-control-plane/envoy/api/v2/listener"
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_api_v2 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	envoy_api_v2_listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
```
Comment (markusthoemmes) [inline pkg/envoy/api/listener_test.go:26]:
```suggestion
	listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
```

File: pkg/generator/caches.go [hunk 8]
```diff
@@ -23,11 +23,11 @@ import (
 	"os"
 	"sync"
 
-	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
-	httpconnmanagerv2 "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	v2 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
```
Comment (markusthoemmes) [inline pkg/generator/caches.go:26]:
```suggestion
	listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
```

File: pkg/generator/caches.go [hunk 9]
```diff
@@ -23,11 +23,11 @@ import (
...
+	route "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	httpconnmanagerv2 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
```
Comment (markusthoemmes) [inline pkg/generator/caches.go:28]:
```suggestion
	hcm "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
```

File: pkg/generator/caches_test.go [hunk 10]
```diff
@@ -21,9 +21,9 @@ import (
 	"sort"
 	"testing"
 
-	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
-	"github.com/envoyproxy/go-control-plane/pkg/resource/v2"
+	v2 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```
Comment (markusthoemmes) [inline pkg/generator/caches_test.go:24]:
```suggestion
	cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```

File: pkg/generator/cluster_cache.go [hunk 11]
```diff
@@ -40,7 +40,7 @@ import (
 	"strings"
 	"time"
 
-	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	v2 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```
Comment (markusthoemmes) [inline pkg/generator/cluster_cache.go:43]:
```suggestion
	cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```

File: pkg/generator/cluster_cache_test.go [hunk 12]
```diff
@@ -21,7 +21,7 @@ import (
 	"testing"
 	"time"
 
-	envoy_api_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	envoy_api_v2 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```
Comment (markusthoemmes) [inline pkg/generator/cluster_cache_test.go:24]:
```suggestion
	cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```

File: pkg/generator/ingress_translator.go [hunk 13]
```diff
@@ -21,9 +21,9 @@ import (
 	"fmt"
 	"time"
 
-	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	endpoint "github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
-	route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
+	v2 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```
Comment (markusthoemmes) [inline pkg/generator/ingress_translator.go:24]:
```suggestion
	cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```

File: pkg/reconciler/ingress/controller.go [hunk 14]
```diff
@@ -19,8 +19,8 @@ package ingress
 import (
 	"context"
 
-	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	xds "github.com/envoyproxy/go-control-plane/pkg/server/v2"
+	v2 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
```
Comment (markusthoemmes) [inline pkg/reconciler/ingress/controller.go:22]:
```suggestion
	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
```

File: pkg/generator/ingress_translator_test.go [hunk 15]
```diff
@@ -20,9 +20,9 @@ import (
 	"testing"
 	"time"
 
-	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	endpoint "github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
-	route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
+	v2 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```
Comment (markusthoemmes) [inline pkg/generator/ingress_translator_test.go:23]:
```suggestion
	cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```

File: pkg/config/ext_authz.go (hunk 2 above)
Comment (jmprusi) [reply pkg/config/ext_authz.go:26]: Oh yes.. I still need to clean all of that.. don't review it yet ;)

File: config/300-gateway.yaml (hunk 1 above)
Comment (jmprusi) [reply config/300-gateway.yaml:39]: bad news.. 1.16 complains about the config due to a bug in envoy.. 1.17 works, I'm just playing around, trying to get the test passing, and then I will try to make it work on 1.16 and 1.14 properly

Comment (jmprusi) [reply config/300-gateway.yaml:39]: I need to dig a little bit more into that issue: https://github.com/envoyproxy/envoy/issues/14272

Comment (markusthoemmes) [reply config/300-gateway.yaml:39]: Gotcha, okay. Might make sense to slap a `[WIP]` on the PR.
